### PR TITLE
Fix non exhaustive match warnings in daml-lf [DPP1281]

### PR DESCRIPTION
**Pull request by mziolekda**
It fixes the warnings of the form
```
daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/EngineTest.scala:2049: warning: match may not be exhaustive.
```
They stem from turning on the `-Xlint:valpattern` scalac flag

### CODE REVIEW

**Comment by remyhaemmerle-da** [comment]
I really would like we continue using relaxed flags for compiling tests.

**Comment by remyhaemmerle-da** [comment]
I already on turn very restrictive flags for non test code, except `daml-lf/api-type-signature` which is own by @S11001001.

**Comment by mziolekda** [comment]
@remyhaemmerle-da 

- Please note that such code will not compile under stricter canton rules
- The way I have rewritten the tests is that the partial patterns are still used, but they are introduced through a `inside` combinator which is a native and clean mechanism within scalatest for exactly that kind of usage. Do you consider that less convenient?

**Comment by S11001001** [comment]
> Please note that such code will not compile under stricter canton rules

I suggest that we shouldn't apply them to this code, then.

**Comment by S11001001** [comment]
> We have to find a compromise.

Since different modules can have different flags, I think we can indefinitely delay ever finding any compromise.

> I prefer more tests a bit less well written that fewer test.

I'm fine with _a lot less_ well written.

> 1. I will argue that inside should be use only on the output of the test, and not for the part that setup the test.

`inside`'s signature itself suggests the opposite by returning the consequent rather than a fixed type like `Unit` or `Assertion`. The only real restriction seems to be that it ought to be evaluated in the dynamic extent of a test execution.

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/LimitsSpec.scala [hunk 1]
```diff
@@ -360,15 +375,16 @@ class LimitsSpec extends AnyWordSpec with Matchers with Inside with TableDrivenP
 
   private val txSeed = crypto.Hash.hashPrivateKey(this.getClass.getCanonicalName)
 
-  private[this] val T = { val Ast.TTyCon(t) = t"Mod:T"; t }
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/LimitsSpec.scala:363]: I would rather use the following definition
```  private[this] val T = 
    Identifier(defaultParserParameters.defaultPackageId, Ref.QualifiedName.assertFromString("Mod:T"))
```

File: daml-lf/interpreter/src/test/scala/com/daml/SBuiltinInterfaceTest.scala [hunk 2]
```diff
@@ -132,7 +136,7 @@ object SBuiltinInterfaceTest {
   private[this] val alice = Ref.Party.assertFromString("Alice")
   private[this] val bob = Ref.Party.assertFromString("Bob")
 
-  import defaultParserParameters.{defaultPackageId => basePkgId}
+  val basePkgId = defaultParserParameters.defaultPackageId
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/test/scala/com/daml/SBuiltinInterfaceTest.scala:139]: what is the issue here ?

File: daml-lf/interpreter/src/test/scala/com/daml/SBuiltinInterfaceTest.scala [hunk 3]
```diff
@@ -267,8 +271,6 @@ object SBuiltinInterfaceTest {
   }
   lazy val compiledBasePkgs = PureCompiledPackages.assertBuild(basePkgs)
 
-  private[lf] val Ast.TTyCon(iouId) = t"'$basePkgId':Mod:Iou"
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/test/scala/com/daml/SBuiltinInterfaceTest.scala:270]:
```
private[lf] val iouId = Ref.Idenfitier.assertFromString("basePkgId':Mod:Iou")
```

File: daml-lf/interpreter/src/test/scala/com/daml/SBuiltinInterfaceTest.scala [hunk 4]
```diff
@@ -302,8 +304,6 @@ object SBuiltinInterfaceTest {
   }
   lazy val compiledExtendedPkgs = PureCompiledPackages.assertBuild(extendedPkgs)
 
-  private val Ast.TTyCon(extraIouId) = t"'$extraPkgId':Mod:Iou"
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/test/scala/com/daml/SBuiltinInterfaceTest.scala:305]: private[lf] val iouId = Ref.Idenfitier.assertFromString("extraPkgId':Mod:Iou")

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/ComparisonSBuiltinTest.scala [hunk 5]
```diff
@@ -416,204 +417,203 @@ class ComparisonSBuiltinTest extends AnyWordSpec with Matchers with TableDrivenP
       def text(s: String) = EPrimLit(PLText(s))
 
       val eitherT = t"Mod:Either"
-      val TTyCon(eitherTyCon) = eitherT
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/ComparisonSBuiltinTest.scala:419]:
```
      val eitherTypeCon = Ref.Identifier(defaultParserParameters.defaultPackageId, Ref.QualifiedName.assertFromString("Mod:Either"))
      val eitherT = TTyCon(t"Mod:Either")
```

**Comment by som-snytt** [comment]
Note `-Xlint:valpattern` just omits the `@unchecked`, for behavior closer to Scala 3, but what warns depends on the compiler.

I just noticed, for example, this code requires `-Xlint:strict-unsealed-patmat,valpattern` in order to warn:
```
  val r = raw"(\d+)".r
  def f(s: String): Int = {
    val r(i) = s
    i.toInt
  }
```
The usual way to say it's OK is to add it back:
```
    val r(i) = s: @unchecked
```
The Scala 3 warning looks like
```
10 |  val D(x) = t
   |             ^
   |        pattern's type D is more specialized than the right hand side expression's type T
   |
   |        If the narrowing is intentional, this can be communicated by adding `: @unchecked` after the expression,
   |        which may result in a MatchError at runtime.
   |        This patch can be rewritten automatically under -rewrite -source 3.2-migration.
```
An egregious example of too much linting in test was adding parens to test methods. On the other hand, every so often, one notices that a test isn't even testing anything, a condition that is non-trivial to detect, but my point is that there are benefits to safety in test.
```
@Test def `test methods are not invoked in code so omit parens in definition`: Unit = ()
```

**Pull request closed without merging**